### DISCUSSION TITLE
Error in lesson 10 (and following question)

### DISCUSSION THREAD

**Discussion by AlbertoSinigaglia** [smartcontractkit/full-blockchain-solidity-course-py#49] [Q&A] [answered]
Hi, in the lesson 10, in the `repay_all` function, you are asking back `amount` tokens.  
However, that is the "value in eth" of the token that your portfolio holds, not the number of tokens that you want to repay back... so instead of:
```python
repayAll(amount, lendingPool, account)
# check using print(getDaiGateway().balanceOf(account)) at the end of the script
```
it should be:
```python
repayAll(Web3.toWei(borrowable_dai, "ether"), lendingPool, account)
```

However at this point, if we check our account balance, we are not even. Infact, our balance will be `99.9 eth` and not the starting `100 eth`... this is because we payed Aave 0.1 ETH in WETH, and Aave is still holding those eth for us (this is also what was asking @chakibhi here [https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/30](https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/30))

How do we get those 0.1ETH back? because those 0.1 are not under some kind of ERC20 token or something else, but just probably in a `mapping(address => eth)` inside the Aave contract

**Comment by PatrickAlphaC** [comment, marked as answer]
Great question.

So we just repaid everything that we had currently borrowed. But even in that split second that we borrowed it, we accrued interest, and that's what you'll see in that `total_debt` varible. However, we don't have any more DAI to pay off that debt. To finish paying it off, we'd have to go buy more DAI first. This is a bit longer to explain, so instead, I have some scripts and code here that can show you how to "buy" more DAI in a forked environment. https://github.com/PatrickAlphaC/defi_py_mix/blob/main/scripts/swap/swap.py

Once you buy more DAI, you can use it to pay the rest of your DAI debt off.


See this answer for the more verbose answer: https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/30#discussioncomment-1358216

**Reply by PatrickAlphaC** [reply]
Let me know if this does/doesn't make sense.

**Reply by AlbertoSinigaglia** [reply]
hi, yes it totally makes sense, however I was looking forward to go back to balance 0 in everything, and buying some DAI won't let me do that.

Since you showed in the video that you can repay with collateral, I was looking for a way to do that, something like:
```python
print("repaying remaining debt using eth")
getAaveWethGateway().repayETH(
    lending_pool,
    min(Web3.toWei(eth_debt, "ether") * 2, account.balance()),
    0,
    account,
    {
        "from": account,
        "value": min(Web3.toWei(eth_debt, "ether") * 2, account.balance())
    }
).wait(1)
print("repayed remaining debt using eth")
```
where `getAaveWethGateway` returns the [aave weth gateway](https://docs.aave.com/developers/the-core-protocol/weth-gateway), that is also the gateway to get back eth from aave... however this code doesn't work it returns this:
```python
  File "brownie/_cli/run.py", line 49, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File "./scripts/borrow.py", line 145, in main
    borrow()
  File "./scripts/borrow.py", line 82, in borrow
    withdrawEthFromAave(totalDebtETH)
  File "./scripts/borrow.py", line 93, in withdrawEthFromAave
    getAaveWethGateway().repayETH(
  File "brownie/network/contract.py", line 1693, in __call__
    return self.transact(*args)
  File "brownie/network/contract.py", line 1566, in transact
    return tx["from"].transfer(
  File "brownie/network/account.py", line 680, in transfer
    receipt._raise_if_reverted(exc)
  File "brownie/network/transaction.py", line 406, in _raise_if_reverted
    raise exc._with_attr(
VirtualMachineError: revert: 15
Trace step 4089, program counter 2041:
  File "contracts/protocol/lendingpool/LendingPool.sol", lines 249-256, in LendingPool.repay:    

        ValidationLogic.validateRepay(
          reserve,
          amount,
          interestRateMode,
          onBehalfOf,
          stableDebt,
          variableDebt
        );
```
that looking [at the code ](https://github.com/aave/protocol-v2/blob/ccb43f026f4ef25ba338073d02dec5c58123da32/contracts/protocol/libraries/helpers/Errors.sol#L42) it is:
```
string public constant VL_NO_DEBT_OF_SELECTED_TYPE = '15'; // 'for repayment of stable debt, the user needs to have stable debt, otherwise, he needs to have variable debt'
```
however, the code doesn't work no matter which debt type I use


I will be very very very grateful if you might have a look at this

**Reply by PatrickAlphaC** [reply]
Yes, I _think_ it actually is still converting to DAI behind the scenes in the UI. I'd have to double-check though.... If you find out, let me know!

Maybe we can ask in the aave discord.
